content/2023-05-03-beware-of-unified-builds.rst

```diff
@@ -0,0 +1,317 @@
+   void bar() {
+     cout << "hello" << std::endl;
+   }
+
+You can use symbols from namespace used from other compilation unit. That's
+exactly the same problem as leaking macro or static definitions: it breaks the
+compilation unit scope.
+
+Putting function implementation in header
+-----------------------------------------
+
+.. code-block:: c++
+
+   // foobar.h
+   #ifndef FOOBAR_H
+   #define FOOBAR_H
+   #include <iostream>
+   void foobar() {
+     std::cout << "hello" << std::endl;
+   }
+   #endif
+
+   // a.cpp
+   #include "foobar.h"
+   void foo() {
+     foobar();
+   }
+
+   // b.cpp
+   #include "foobar.h"
+   void bar() {
+     foobar();
+   }
+
+As each header is only included once, you can put your function definition in
+your header. Easy!
+
+Putting constant initializer in implementation
+----------------------------------------------
+
+.. code-block:: c++
+
+   // foobar.h
+   #ifndef FOOBAR_H
+   #define FOOBAR_H
+   struct foo {
+   static const int VALUE;
+   };
+   #endif
+
+   // a.cpp
+   #include "foobar.h"
+   const int foo::VALUE = 1;
+
+   // b.cpp
+   #include "foobar.h"
+   static_assert(foo::VALUE == 1, "ok");
+
+The constant expression lacks its initializer.
+
+Functions with invalid declaration that gets silent
+---------------------------------------------------
+
+
+.. code-block:: c++
+
+   // foobar.h
+   #ifndef FOOBAR_H
+   #define FOOBAR_H
+   void foo(int * ptr);
+   #endif
+
+   // a.cpp
+   #include "foobar.h"
+   void foo(const int * ptr) {
+   }
+
+   // b.cpp
+   #include "foobar.h"
+   void bar(const int * ptr) {
+       return foo(ptr);
+   }
+
+Invalid forward declaration but who cares, when the definition can be found and
+the compiler doesn't warn about unused forward declaration?
+
+Unexpected positive aspect: more warnings
+-----------------------------------------
+
+.. code-block:: c++
+
+   // a.cpp
+   static int foo = 0;
+
+   // b.cpp
+   #include "a.cpp"
+
+Compiling ``a.cpp`` yields an unused warning, but not compiling ``b.cpp``. So
+hybrid builds relying on ``#including`` multiple sources actually decrease the
+warning level.
+
+Headers without guard
+---------------------
+
+.. code-block:: c++
+
+   // foobar.h
+   struct Foo {};
+
+   // a.cpp
+   #include "foobar.h"
+   void a(Foo&) {}
+
+   // b.cpp
+   void b(Foo&) {}
+
+Fixing the include for ``b.cpp`` leads to type redefinition because the header
+is not guarded.
+
+
+About the Firefox codebase
+==========================
+
+The removal of ``REQUIRES_UNIFIED_BUILD`` across the Firefox codebase was
+tracked under https://bugzilla.mozilla.org/show_bug.cgi?id=1626530. Since I
+focused on this, I've landed more than 150 commits, modifying more than 800
+sources files. And it's now done, no more hard requirement of unified build,
+back to a normal situation.
+
+Was it worth the effort? It prevents bad coding practice, the codebase is now
+more friendly to static analyzers which didn't now about unified builds, so yes,
```

Comment:
`which` implies there are static analyzer that know about unified builds and works well with them, I don't think that's the case.

Split the sentence into two and the question / answer so it reads better:

Was it worth the effort? Yes: it prevents bad coding practice**s** (plural), and static analysis is now more useful compared to what it could do with unified builds.